### PR TITLE
add recursive env var to config

### PR DESCRIPTION
**Pull request by ornoone**
imprements the nested resolution of env var in yaml config.

this changes is related to issue #510

it add the feature to parse correctly default value in environment values. 

the feature require extra package __regex__ to accept the recursive (?2) in the regex.

the new behavior is enabled only if this package is installed. if not, the old non recursive regex is used.

### CODE REVIEW

**Comment by ornoone** [comment]
@mattbennett travis seem to complains about spelling on files I did not changes... 

content of spelling/output.txt: 
```
installation.rst:21: (setuptools) 
testing.rst:81: (setuptools) 
```
what can I do to fix it on my branch ?

**Comment by mattbennett** [comment]
@ornoone I don't know what that's suddenly started failing, but adding `setuptools` to `docs/spelling_wordlist.txt` will fix it.

File: docs/cli.rst [hunk 1]
```diff
@@ -73,6 +74,15 @@ If you need to quote the values in your YAML file, the explicit ``!env_var`` res
     # foobar.yaml
     AMQP_URI: !env_var "pyamqp://${RABBITMQ_USER:guest}:${RABBITMQ_PASSWORD:password}@${RABBITMQ_HOST:localhost}"
 
+You can provide many level of generic default value
+
+.. code-block:: yaml
+
+	# foobar.yaml
+	AMQP_URI: ${AMQP_URI:pyamqp://${RABBITMQ_USER:guest}:${RABBITMQ_PASSWORD:password}@${RABBITMQ_HOST:localhost}}
```
Comment (kooba) [inline docs/cli.rst]: Looks like you used `tab` instead of 4 spaces as this file expects.

File: docs/cli.rst [hunk 2]
```diff
@@ -50,9 +50,10 @@ Config values can be read via the built-in :ref:`config_dependency_provider` dep
 
 Environment variable substitution
 ---------------------------------
-YAML configuration files have basic support for environment variables.
+YAML configuration files have support for environment variables.
 You can use bash style syntax: ``${ENV_VAR}``
 Optionally you can provide default values ``${ENV_VAR:default_value}``
+default values can contains environment variables recursively ``${ENV_VAR:default_${OTHER_ENV_VAR:value}}`` (note: this feature require regex package)
```
Comment (kooba) [inline docs/cli.rst]: This seems to be a new sentence. Maybe we should start it with capital letter and end with period? Looks like 2 sentences above yours missing period as well.

File: tox.ini [hunk 3]
```diff
@@ -23,6 +23,7 @@ deps =
     pinned: six==1.10.0
     pinned: werkzeug==0.11.11
     pinned: wrapt==1.10.8
+    pinned: regex==2018.2.21
```
Comment (kooba) [inline tox.ini]: Not 100% sure how this works, but was this supposed to be `extra: regex==2018.2.21`?

**Comment by ornoone** [comment]
I finally passed all test. to fix coverage 100%, i added some ``# pragma: no cover`` to code specific to the presence or not of the package regex.

File: docs/cli.rst [hunk 4]
```diff
@@ -50,9 +50,10 @@ Config values can be read via the built-in :ref:`config_dependency_provider` dep
 
 Environment variable substitution
 ---------------------------------
-YAML configuration files have basic support for environment variables.
-You can use bash style syntax: ``${ENV_VAR}``
-Optionally you can provide default values ``${ENV_VAR:default_value}``
+YAML configuration files have support for environment variables.
+You can use bash style syntax: ``${ENV_VAR}``.
+Optionally you can provide default values ``${ENV_VAR:default_value}``.
+Default values can contains environment variables recursively ``${ENV_VAR:default_${OTHER_ENV_VAR:value}}`` (note: this feature require regex package).
```
Comment (kooba) [inline docs/cli.rst]: `this feature requires regex package`

File: docs/cli.rst [hunk 5]
```diff
@@ -73,6 +74,15 @@ If you need to quote the values in your YAML file, the explicit ``!env_var`` res
     # foobar.yaml
     AMQP_URI: !env_var "pyamqp://${RABBITMQ_USER:guest}:${RABBITMQ_PASSWORD:password}@${RABBITMQ_HOST:localhost}"
 
+You can provide many level of generic default value
```
Comment (kooba) [inline docs/cli.rst]: `many levels of default values`

File: docs/cli.rst [hunk 6]
```diff
@@ -73,6 +74,15 @@ If you need to quote the values in your YAML file, the explicit ``!env_var`` res
...
+
+.. code-block:: yaml
+
+    # foobar.yaml
+    AMQP_URI: ${AMQP_URI:pyamqp://${RABBITMQ_USER:guest}:${RABBITMQ_PASSWORD:password}@${RABBITMQ_HOST:localhost}}
+
+this config accept AMQP_URI as environment variable, in this case RABBITMQ_* variables will not be used.
```
Comment (kooba) [inline docs/cli.rst]: `this config accepts AMQP_URI as an environment variable, if provided RABBITMQ_* nested variables will not be used.`

File: docs/cli.rst [hunk 7]
```diff
@@ -85,6 +95,15 @@ The environment variable value is interpreted as YAML, so it is possible to use
 
     $ A_LIST_OF_THINGS=[A,B,C] nameko run --config ./foobar.yaml <module>[:<ServiceClass>]
 
+the parser for environment variable will pair all bracket.
```
Comment (kooba) [inline docs/cli.rst]: `the parser for environment variables will pair all brackets.`

**Comment by kooba** [comment]
Thank you @ornoone looks great. Just a few more documentation fixes and I think it will be good to go.

**Comment by kooba** [comment]
@mattbennett @davidszotten this looks good to me. This change in combination with global config from https://github.com/nameko/nameko/pull/520 will give users quite a flexibility.

File: test/cli/test_main.py [hunk 8]
```diff
@@ -237,3 +298,100 @@ def test_cannot_recurse(self):
...
+            {"INDICE": 'b'},
+            {"FOO": 'val_${INDICE:1}'},
+        ),
+        # recursive env requiring data
+        (
+            """
+            FOO: ${FOO:${INDICE}_val}
+            """,
+            {"INDICE": 'b'},
+            {"FOO": '${INDICE_val}'},
```
Comment (mattbennett) [inline test/cli/test_main.py]: Shouldn't the expected value here be `${INDICE}_val`? Note the extra `}`

File: test/cli/test_main.py [hunk 9]
```diff
@@ -237,3 +298,100 @@ def test_cannot_recurse(self):
...
+        ),
+
+    ])
+    @pytest.mark.skipif(has_regex_module,
+                        reason='default behavior if no regex module')
+    def test_unhandled_recurtion(
```
Comment (mattbennett) [inline test/cli/test_main.py]: s/recurtion/recursion

File: test/cli/test_main.py (hunk 8 above)
Comment (ornoone) [reply test/cli/test_main.py]: the ENV_VAR_MATCHER regex behavior is to match until the first ``}``. this split the given config into: 
``${FOO:${INDICE}``**+**``_val}``
with ``FOO`` having default value of ``${INDICE`` concatenated with raw string ``_val}``

Comment (mattbennett) [reply test/cli/test_main.py]: Is this desirable behaviour though?

Actually I would expect this case to evaluate as

```
FOO: b_val
```

Comment (mattbennett) [reply test/cli/test_main.py]: Comparing this case to the one above in `test_environment_vars_recursive_in_config`:

```
# test_environment_vars_recursive_in_config:
# recursive env requiring data
(
    """
    FOO: ${FOO:val_${INDICE}}
    """,
    {"INDICE": 'b'},
    {"FOO": 'val_b'},  # << expected value "val_b"
),
```

```
# test_unhandled_recursion:
# recursive env requiring data
(
     """
     FOO: ${FOO:${INDICE}_val}
     """,
     {"INDICE": 'b'},
     {"FOO": '${INDICE_val}'},   # << expected value should be "b_val"
)
```

What's the difference?

Comment (ornoone) [reply test/cli/test_main.py]: I have added test for both old and new behaviors. the behaviors described is for the old behavior ``@pytest.mark.skipif(has_regex_module,`` => don't run if we are in the env with regex, so this run only without regex, and the old behavior is expected.

=> the difference is in the ``@pytest.mark.skipif(==>no<==t has_regex_module,``

Comment (mattbennett) [reply test/cli/test_main.py]: Ahh, I understand now. Sorry for the confusion

**Pull request merged**